Add instructions to write custom checks (#35)

mstoetzer opened the pull request:
Add instruction to extend custom checks with scope and remediation message

@viswajithiii @gaurav-nelson 

Fixes #35

kreamkorokke commented on docs/configuring-kubelinter.md:
````diff
@@ -62,3 +62,59 @@ example,
+          - DeploymentLike
+  ```
+
+  At the time of this writing, `objectKinds` can be either `DeploymentLike` or `Any`.
````
In fact we also support `Service` as well. Maybe instead of enumerating all the object kinds here, we can insert a link to this folder https://github.com/stackrox/kube-linter/tree/main/internal/objectkinds and say all supported object kinds can be found here?

mstoetzer replied:
Sound muchs better, yes :+1:

kreamkorokke commented:
Also I just merged a fix which should resolve the circle image job failure, could you rebase and give it a try? Thank you!

kreamkorokke commented on docs/configuring-kubelinter.md:
```diff
@@ -62,3 +62,59 @@ example,
+
+## Run custom checks
+
+You can write custom checks that are based on existing templates.
```
Can we also insert a link to all the existing templates here? And also please mention that params definitions are also taken from the respective templates.

gaurav-nelson replied:
```suggestion
You can write custom checks based on existing [templates](generated/templates.md).
```

gaurav-nelson commented on docs/configuring-kubelinter.md:
```diff
@@ -62,3 +62,59 @@ example,
+
+You can write custom checks that are based on existing templates.
+
+For example,
+- To be sure that an annotation exists, you can add the following:
```
```suggestion
For example,
- To make sure that an annotation exists, you can add the [`required-annotation`](generated/templates?id=required-annotation) template:
```

gaurav-nelson commented on docs/configuring-kubelinter.md:
````diff
@@ -62,3 +62,59 @@ example,
+        key: company.io/responsible
+  ```
+
+  It uses the existing `required-annotation` template and adds the check for your custom keys.
````
```suggestion
```

gaurav-nelson commented on docs/configuring-kubelinter.md:
```diff
@@ -62,3 +62,59 @@ example,
+
+  It uses the existing `required-annotation` template and adds the check for your custom keys.
+
+- To be sure that a specific label exists, you can add the following:
```
```suggestion
- To make sure that a specific label exists, you can use the [`required-label`](generated/templates?id=required-label) template:
```

gaurav-nelson commented on docs/configuring-kubelinter.md:
````diff
@@ -62,3 +62,59 @@ example,
+        key: company.io/release
+  ```
+
+  It uses the existing `required-label` template.
````
```suggestion
```

gaurav-nelson commented on docs/configuring-kubelinter.md:
```diff
@@ -62,3 +62,59 @@ example,
+
+  It uses the existing `required-label` template.
+
+### Extending custom checks
```
```suggestion
### Extend custom checks
```

To be inline with other headings

gaurav-nelson commented on docs/configuring-kubelinter.md:
```diff
@@ -62,3 +62,59 @@ example,
+
+### Extending custom checks
+
+There may be situations where custom checks need to be scoped to specific Kubernetes objects. To make custom checks more user friendly, a message can be added to the output of the check. These can be added by using `scope` or `remediation` respectively.
```
```suggestion
With custom checks, you can control the checks to run only on specific Kubernetes object types (such as services or deployments). You can also modify the remediation message you get when your custom check fails. For example,
```

gaurav-nelson commented on docs/configuring-kubelinter.md:
```diff
@@ -62,3 +62,59 @@ example,
+There may be situations where custom checks need to be scoped to specific Kubernetes objects. To make custom checks more user friendly, a message can be added to the output of the check. These can be added by using `scope` or `remediation` respectively.
+
+For example,
+- To be sure that only Deployments have set a required annotation, you can add the following:
```
```suggestion
- To make sure that a specific annotation exists on all deployments, you can use the [`required-label`](generated/templates?id=required-label) template and specify a Kubernetes objects `scope`:
```

gaurav-nelson commented on docs/configuring-kubelinter.md:
````diff
@@ -62,3 +62,59 @@ example,
+          - DeploymentLike
+  ```
+
+  At the time of this writing, `objectKinds` can be either `DeploymentLike` or `Any`.
````
```suggestion
  For details about `objectKinds` that KubeLinter support, see https://github.com/stackrox/kube-linter/tree/main/internal/objectkinds
```

gaurav-nelson commented on docs/configuring-kubelinter.md:
```diff
@@ -62,3 +62,59 @@ example,
+
+  At the time of this writing, `objectKinds` can be either `DeploymentLike` or `Any`.
+
+- To be more user friendly it is recommended to add a message to the check, that describes, what to do when this check fails:
```
```suggestion
- Use `remediation` to include a remediation message that users get when your custom check fails:
```

gaurav-nelson reviewed (changes requested):
Thanks @mstoetzer for the PR, I made a few suggestions that improves the language.

mstoetzer commented:
> Thanks @mstoetzer for the PR, I made a few suggestions that improves the language.

Thanks @kreamkorokke and @gaurav-nelson for your efforts. I rebased and updated the commit according to your suggestions. The language sounds much better now :)

gaurav-nelson commented on docs/configuring-kubelinter.md:
```diff
@@ -62,3 +62,55 @@ example,
+
+## Run custom checks
+
+You can write custom checks based on existing [templates](generated/templates.md). All `params` can be taken from the respective templates.
```
```suggestion
You can write custom checks based on existing [templates](generated/templates.md). Every template description includes details about the parameters (`params`) you can use along with that template.
```

gaurav-nelson reviewed (approved):
LGTM

viswajithiii commented:
This closes #35

viswajithiii commented:
Thanks again for the contribution @mstoetzer!
